pulsar-common/src/main/java/org/apache/pulsar/common/protocol/PulsarDecoder.java

```diff
@@ -121,6 +121,8 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
             int cmdSize = (int) buffer.readUnsignedInt();
             cmd.parseFrom(buffer, cmdSize);
 
+            checkRateLimit(cmd);
```

Comment:
this method `checkRateLimit` is very specific in an abstract class. One possible solution would be to remove this here and instead change the `messageReceived` method to accept `cmd` as a parameter. The default implementation would just delegate to `messageReceived()` without parameters. This would keep the abstract class clean while serving the changes you are intended to make.

Reply:
Good suggestion, improved